8300575: JVMTI support when using alternative virtual thread implementation

pchilano opened the pull request:
Please review the following change. Some of the JVMTI functions had to be modified since they are not supported by the specs on virtual threads (StopThread, RunAgentThread, GetCurrentThreadCpuTime, GetThreadCpuTime) or shouldn't include virtual threads in the results (GetAllThreads, GetAllStackTraces, GetThreadGroupChildren). Others were modified because they should work on virtual thread but they weren't (SuspendAllVirtualThreads/ResumeAllVirtualThreads). Also support for VirtualThreadStart/VirtualThreadEnd events was added.

There were a total of 71 tests under serviceability/jvmti/ that had the vm.continuations requirement. Of those, 24 where under serviceability/jvmti/vthread/. I was able to remove that requirement from 50 tests. Most of those tests were already passing for the alternative vthread implementation just by removing the check in jni_GetEnv for VMContinuations.
From the other 21 tests, 20 still fail with the changes either because they actually test continuations or because of different assumptions in the tests that don't hold true for the alternative vthread implementation. So for those I left the vm.continuations requirement untouched (from those 20 tests there are actually 4 tests from the thread/GetStackTrace/ family that are passing because of a bug in the test, but I can fix that in another RFE). The other remaining test is vthread/GetSetLocalTest/GetSetLocalTest.java which I see is problemlisted.

Regarding variable _is_bound_vthread, although it's handy as a cache to avoid repeating the check, I mainly added it to avoid transitioning for GetCurrentThreadCpuTime (we are native and cannot access oops).

I added new test BoundVThreadTest.java which just checks for the unsupported/supported behavior mentioned previously. For some extra basic coverage I also added a new run with -XX:-VMContinuations on tests inside the serviceability/jvmti/vthread/ directory that don't require vm.continuations anymore. I could also add that for all the other tests in serviceability/jvmti/ for which I removed the vm.continuations requirement.

I run the patch through mach5 tiers 1-6 plus some extra local runs on the relevant tests.

Thanks,
Patricio

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8300575](https://bugs.openjdk.org/browse/JDK-8300575): JVMTI support when using alternative virtual thread implementation


### Reviewers
 * [Leonid Mesnik](https://openjdk.org/census#lmesnik) (@lmesnik - **Reviewer**) ⚠️ Review applies to [99cf204c](https://git.openjdk.org/jdk/pull/12512/files/99cf204cf6b07cfc83d907f48ee7a95ffd69c7d4)
 * [Serguei Spitsyn](https://openjdk.org/census#sspitsyn) (@sspitsyn - **Reviewer**)
 * [Alan Bateman](https://openjdk.org/census#alanb) (@AlanBateman - **Reviewer**)


### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk pull/12512/head:pull/12512` \
`$ git checkout pull/12512`

Update a local copy of the PR: \
`$ git checkout pull/12512` \
`$ git pull https://git.openjdk.org/jdk pull/12512/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 12512`

View PR using the GUI difftool: \
`$ git pr show -t 12512`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/12512.diff">https://git.openjdk.org/jdk/pull/12512.diff</a>

</details>

pchilano commented:
/label remove core-libs

AlanBateman commented on src/hotspot/share/classfile/vmClassMacros.hpp:
```diff
@@ -91,6 +91,7 @@
   do_klass(ThreadGroup_klass,                           java_lang_ThreadGroup                                 ) \
   do_klass(BasicVirtualThread_klass,                    java_lang_BaseVirtualThread                           ) \
```
It should be BaseVirtualThread_klass rather BasicVirtualThread_klass (probably my fault when adding the alt implementation).

pchilano replied:
Fixed.

AlanBateman commented:
I see that JVM_GetAllThreads will now returned the filtered list (as an array). That looks okay. There's another part to this in the ThreadMXBean implementation where the filtering is done in the Java implementation. We can look at that in a follow-up issue, to avoid adding more to this PR.

On tests, there are a few other tests that can be changed to drop `@requires vm.continuations`, e.g. the jshell tests that run with --enable-preview should no longer need the `@requires`. There are also a few JDI and at least one AppCDS test. A search will find them. The tests re-running with -XX:-VMContinuations is fine but it does mean they will run twice on the ports without an implementation. The test runtime/jni/IsVirtualThread/IsVirtualThread.java is an example where it only runs once on these ports, at the cost of additional `@test` tag.

AlanBateman commented:
> Regarding variable _is_bound_vthread, although it's handy as a cache to avoid repeating the check, I mainly added it to avoid transitioning for GetCurrentThreadCpuTime (we are native and cannot access oops).

The timer functions need re-visiting to allow an implementation support these functions on virtual threads. Same for StopThread which needs a spec change to minimally support this function on a thread that is suspended at a breakpoint or single step event. So if we sort out these spec issues then it looks like JavaThread:: _is_bound_vthread could go away and there would be less coupling with the alt implementation.

pchilano commented:
> On tests, there are a few other tests that can be changed to drop `@requires vm.continuations`, e.g. the jshell tests that run with --enable-preview should no longer need the `@requires`. There are also a few JDI and at least one AppCDS test. A search will find them. 
>
I removed the vm.continuations requirement on those tests (only SuspendAfterDeath.java and RedefineRunningMethods_Shared.java were actually creating a JVMTI environment though).

>The tests re-running with -XX:-VMContinuations is fine but it does mean they will run twice on the ports without an implementation. The test runtime/jni/IsVirtualThread/IsVirtualThread.java is an example where it only runs once on these ports, at the cost of additional `@test` tag.
>
I see, I removed the extra run and created another @test tag for that with `@requires vm.continuations`, like it is done in IsVirtualThread.java.

pchilano commented:
Also fixed a bug in the description. PopFrame and ForceEarlyReturnXXX don't actually need to return UNSUPPORTED_OPERATION (optionally supported by the specs), but when looking at jvmtiEnv.cpp, I saw the "// No support for virtual threads (yet)" comment, so I just changed them to return UNSUPPORTED_OPERATION for the alternative implementation too to match the behavior with normal virtual threads. But I could revert that if agreed. In fact I left SetLocalXXX unmodified where currently the implementation has different support on virtual threads than platform threads.

AlanBateman commented:
The changes in the updates look good. I think the main thing to decide on is whether JavaThread:: _is_bound_vthread is needed or not. If GetCurrentThreadCpuTime is changed to allow an implementation support this function then it looks like the additional field would not be needed.

lmesnik commented on test/hotspot/jtreg/serviceability/jvmti/vthread/BoundVThreadTest/BoundVThreadTest.java:
```diff
@@ -0,0 +1,73 @@
+/**
+ * @test
+ * @summary Verifies correct JVMTI behavior for BoundVirtualThreads.
+ * @compile --enable-preview -source ${jdk.version} BoundVThreadTest.java
```
You could use @enablePreview instead.

pchilano replied:
Changed.

lmesnik reviewed (approved):
I've left one comment. Otherwise test changes looks good to me.

pchilano commented:
> The changes in the updates look good. I think the main thing to decide on is whether JavaThread:: _is_bound_vthread is needed or not. If GetCurrentThreadCpuTime is changed to allow an implementation support this function then it looks like the additional field would not be needed.
>
Great, thanks for looking at this Alan. I'll remove _is_bound_vthread if we change the specs for GetCurrentThreadCpuTime then.

pchilano commented:
> I've left one comment. Otherwise test changes looks good to me.
>
Thanks for the review Leonid!

sspitsyn commented on src/hotspot/share/prims/jvmtiEnv.cpp:
```diff
@@ -1042,9 +1039,10 @@ JvmtiEnv::SuspendAllVirtualThreads(jint except_count, const jthread* except_list
+          ((java_lang_VirtualThread::is_instance(vt_oop) &&
+            JvmtiEnvBase::is_vthread_alive(vt_oop) &&
+            !JvmtiVTSuspender::is_vthread_suspended(vt_oop)) ||
+            java_thread->is_bound_vthread()) &&
```
This condition does not look correct.
I'd expect it to be:
 ```
         ((java_lang_VirtualThread::is_instance(vt_oop) &&
           JvmtiEnvBase::is_vthread_alive(vt_oop)) ||
          java_thread->is_bound_vthread()) &&
          !JvmtiVTSuspender::is_vthread_suspended(vt_oop) &&
```
It is important to check for `!JvmtiVTSuspender::is_vthread_suspended(vt_oop)` for `bound_vthread` as well.
The same issue is in the `JvmtiEnv::ResumeAllVirtualThreads`.

pchilano replied:
The thing is that JvmtiVTSuspender::is_vthread_suspended() checks _suspended_list/not_suspended_list which are not set with single suspend/resume functions(SuspendThread/SuspendThreadList/ResumeThread/ResumeThreadList) on a BoundVirtualThread. For single suspend/resume on a BoundVirtualThread we go through the same path as for platform threads. In any case, the suspend_thread() call will just return JVMTI_ERROR_THREAD_SUSPENDED if a thread is already suspended, but if you want to avoid the call altogether I can add the equivalent check which would be !java_thread->is_suspended().

Now, I realized that although single-suspends will not change _suspended_list/not_suspended_list for a BoundVirtualThread, SuspendAllVirtualThreads/ResumeAllVirtualThreads can if a BoundVirtualThread is part of the exception list. In the SuspendAllVirtualThreads case for example, when restoring the resume state, the call to JvmtiVTSuspender::is_vthread_suspended() would return true (we are already in SR_all mode and the BoundVirtualThread will not be in _not_suspended_list), meaning the BoundVirtualThread will be added to the not_suspended_list. 
So we have to decide if we want to keep track of BoundVirtualThreads on these lists or not. Given that I see JvmtiVTSuspender::is_vthread_suspended() is always called on a VirtualThread the easiest thing would be to not track them, and just add a check in SuspendAllVirtualThreads/ResumeAllVirtualThreads where the elist is populated so only VirtualThreads are added. What do you think?

sspitsyn replied:
The `SuspendAllVirtualThreads` spec has this statement:
```
  Suspend all virtual threads except those in the exception list.
  Virtual threads that are currently suspended do not change state.
```
And the `ResumeAllVirtualThreads` spec has this statement:
```
  Resume all virtual threads except those in the exception list.
  Virtual threads that are currently resumed do not change state
```
So that, `SuspendAllVirtualThreads` should do nothing with the already suspended virtual threads.
And `ResumendAllVirtualThreads` should do nothing with the already resumed virtual threads.
It does not depend on the except_list.
It is better to avoid suspending already suspended threads and resume already resumed.
So, at least a check for `!java_thread->is_suspended()` is needed.
Did you observe any JVMTI or JDI suspend/resume tests failing?

> Now, I realized that although single-suspends will not change _suspended_list/not_suspended_list for a BoundVirtualThread, SuspendAllVirtualThreads/ResumeAllVirtualThreads can if a BoundVirtualThread is part of the exception list. In the SuspendAllVirtualThreads case for example, when restoring the resume state, the call to JvmtiVTSuspender::is_vthread_suspended() would return true (we are already in SR_all mode and the BoundVirtualThread will not be in _not_suspended_list), meaning the BoundVirtualThread will be added to the not_suspended_list.
> So we have to decide if we want to keep track of BoundVirtualThreads on these lists or not. Given that I see JvmtiVTSuspender::is_vthread_suspended() is always called on a VirtualThread the easiest thing would be to not track them, and just add a check in SuspendAllVirtualThreads/ResumeAllVirtualThreads where the elist is populated so only VirtualThreads are added. What do you think?

I feel that following the VirtualThread's implementation code path would be more consistent, unified and free of surprises. But I also see that following the platform threads code path looks pretty simple. So, it can be okay.
So, for this approach, I think, the code fragments which collect and restore resumed/suspended state for virtual threads need to be bypassed by the BoundVirtualThreads implementation. Please, let me know if it does not work by any reason. We also will have to rely on the testing here.

pchilano replied:
> The `SuspendAllVirtualThreads` spec has this statement:
> 
> ```
>   Suspend all virtual threads except those in the exception list.
>   Virtual threads that are currently suspended do not change state.
> ```
> 
> And the `ResumeAllVirtualThreads` spec has this statement:
> 
> ```
>   Resume all virtual threads except those in the exception list.
>   Virtual threads that are currently resumed do not change state
> ```
> 
> So that, `SuspendAllVirtualThreads` should do nothing with the already suspended virtual threads. And `ResumeAllVirtualThreads` should do nothing with the already resumed virtual threads. It does not depend on the except_list. It is better to avoid suspending already suspended threads and resuming already resumed. So, at least a check for `!java_thread->is_suspended()` is needed. 
>
Yes, the elist comment was related to the other issue. The reason why I didn't explicitly added this check is because it will we done in the suspend/resume call already (while holding the handshake lock which serializes the suspend/resume requests). So inside the suspend handshake for example either the target is suspended, so nothing is done, or the target is not suspended and we suspend it. Is there an actual issue I'm missing or is this to make a fast check and avoid the call? In any case I added the check. Also I see the specs for the single suspend/resume functions say the same thing about not changing the state and, for example, in SuspendThreadList we are also not making an explicit check before the call to suspend_thread(). So if there is an actual issue here, don't we need to add that check for them too?

>Did you observe any JVMTI or JDI suspend/resume tests failing?
>
No. I'll retest the patch again though once I resolve all comments.
 
> > Now, I realized that although single-suspends will not change _suspended_list/not_suspended_list for a BoundVirtualThread, SuspendAllVirtualThreads/ResumeAllVirtualThreads can if a BoundVirtualThread is part of the exception list. In the SuspendAllVirtualThreads case for example, when restoring the resume state, the call to JvmtiVTSuspender::is_vthread_suspended() would return true (we are already in SR_all mode and the BoundVirtualThread will not be in _not_suspended_list), meaning the BoundVirtualThread will be added to the not_suspended_list.
> > So we have to decide if we want to keep track of BoundVirtualThreads on these lists or not. Given that I see JvmtiVTSuspender::is_vthread_suspended() is always called on a VirtualThread the easiest thing would be to not track them, and just add a check in SuspendAllVirtualThreads/ResumeAllVirtualThreads where the elist is populated so only VirtualThreads are added. What do you think?
> 
> I feel that following the VirtualThread's implementation code path would be more consistent, unified and free of surprises. But I also see that following the platform threads code path looks pretty simple. So, it can be okay. So, for this approach, I think, the code fragments which collect and restore resumed/suspended state for virtual threads need to be bypassed by the BoundVirtualThreads implementation, so the `_suspended_list`/`not_suspended_list` won't be impacted. Please, let me know if it does not work by any reason. We also will have to rely on the testing here.
>
Sounds good. Yes, just following the platform thread code path is easier. Plus the other functions that are supported on virtual threads already have a branch for platform thread vs virtual thread, and the BoundVirtualThread already takes the platform thread path. Last commit is a fix for this issue in SuspendAllVirtualThreads/ResumeAllVirtualThreads, please take a look.

sspitsyn replied:
> Yes, the elist comment was related to the other issue. The reason why I didn't explicitly added this check is because it will we done in the suspend/resume call already (while holding the handshake lock which serializes the suspend/resume requests). So inside the suspend handshake for example either the target is suspended, so nothing is done, or the target is not suspended and we suspend it. Is there an actual issue I'm missing or is this to make a fast check and avoid the call? In any case I added the check. Also I see the specs for the single suspend/resume functions say the same thing about not changing the state and, for example, in SuspendThreadList we are also not making an explicit check before the call to suspend_thread(). So if there is an actual issue here, don't we need to add that check for them too?

I agree with your analysis. And you are write on the `SuspendThreadList` as well.
The `SuspendThreadList` also return an error code of each suspend in the result array.
It'd be nice to add a check there as well to make it consistent but I'd suggest to be conservative here and leave it as it is now.

sspitsyn reviewed (commented):
Patricio,
The fix looks pretty solid to me.
I've already posted one comment and will post a couple of formatting nits.
I agree with Alan, it'd be nice to get rid of `is_bound_vthread` if possible.
I will make one more pass after your update.
Also, I'd address the JVMTI spec issue (mentioned by Alan) in a separate CR that needs to go with a CSR.
I'll file an RFE or spec bug on it.
Thanks,
Serguei

sspitsyn commented on src/hotspot/share/prims/jvmtiEnvBase.cpp:
```diff
@@ -1583,7 +1583,8 @@ JvmtiEnvBase::suspend_thread(oop thread_oop, JavaThread* java_thread, bool singl
   if (!is_passive_cthread) {
-    assert(single_suspend || is_virtual, "SuspendAllVirtualThreads should never suspend non-virtual threads");
+    assert(thread_h() != nullptr, "sanity check");
+    assert(single_suspend || thread_h()->is_a(vmClasses::BaseVirtualThread_klass()), "SuspendAllVirtualThreads should never suspend non-virtual threads");
```
Nit: It is better to split this line by placing assert message on a separate line.

pchilano replied:
Fixed.

sspitsyn commented on src/hotspot/share/prims/jvmtiEnvBase.cpp:
```diff
@@ -1643,7 +1644,8 @@ JvmtiEnvBase::resume_thread(oop thread_oop, JavaThread* java_thread, bool single
   if (!is_passive_cthread) {
-    assert(single_resume || is_virtual, "ResumeAllVirtualThreads should never resume non-virtual threads");
+    assert(thread_h() != nullptr, "sanity check");
+    assert(single_resume || thread_h()->is_a(vmClasses::BaseVirtualThread_klass()), "ResumeAllVirtualThreads should never resume non-virtual threads");
```
Nit: It is better to split this line by placing assert message on a separate line.

pchilano replied:
Fixed.

pchilano commented:
> Patricio, The fix looks pretty solid to me. I've already posted one comment and will post a couple of formatting nits. I agree with Alan, it'd be nice to get rid of `is_bound_vthread` if possible. I will make one more pass after your update. Also, I'd address the JVMTI spec issue (mentioned by Alan) in a separate CR that needs to go with a CSR. I'll file an RFE or spec bug on it. Thanks, Serguei
>
Thanks for looking at this Serguei.

sspitsyn commented on test/hotspot/jtreg/serviceability/jvmti/vthread/BoundVThreadTest/libBoundVThreadTest.cpp:
```diff
@@ -0,0 +1,310 @@
+static jvmtiEnv *jvmti = nullptr;
+static int vthread_start_count = 0;
+static int vthread_end_count = 0;
+static bool passed = JNI_TRUE;
```
Nit: It is better to rename: passed => status.

pchilano replied:
Fixed.

sspitsyn reviewed (approved):
The fix looks good to me. Thank you for the update.
I'm not sure if you wanted to get rid of `is_bound_vthread` though.
Just want to know you intention.
Thanks,
Serguei

pchilano commented:
> The fix looks good to me. Thank you for the update. I'm not sure if you wanted to get rid of `is_bound_vthread` though. Just want to know you intention. Thanks, Serguei
>
Great. Yes, I just removed is_bound_vthread, please take a look when you have time. I'll also re-run all testing again before integration. 
Thanks for the review Serguei!

AlanBateman commented:
Thanks for dropping is_bound_vthread, I think it's much better now. What would you think about changing it to test the thread type is_a vmClasses::BaseVirtualThread_klass() so the VM doesn't need to know about ThreadBuilders$BoundVirtualThread.

pchilano commented:
> Thanks for dropping is_bound_vthread, I think it's much better now. What would you think about changing it to test the thread type is_a vmClasses::BaseVirtualThread_klass() so the VM doesn't need to know about ThreadBuilders$BoundVirtualThread.
>
Yes, we could do that. Although I think that checking for BoundVirtualThread_klass is more descriptive of what we are trying to do and will be less confusing when reading the code. How about checking for `(!VMContinuations && x->is_a(vmClasses::BaseVirtualThread_klass()))`?

sspitsyn reviewed (approved):
Updates look good to me.
Thanks,
Serguei

AlanBateman commented:
> Yes, we could do that. Although I think that checking for BoundVirtualThread_klass is more descriptive of what we are trying to do and will be less confusing when reading the code. How about checking for `(!VMContinuations && x->is_a(vmClasses::BaseVirtualThread_klass()))`?

We can go with what you have for now and re-visit again if it becomes an issue. The main thing I had hoped is that the VM won't get too coupled to BoundVirtualThread as that may move or be refactored soon.

pchilano commented:
Thanks for the reviews @AlanBateman, @lmesnik and @sspitsyn!

pchilano commented:
/integrate
